### PR TITLE
[PyOV] Add more ov.Model constructors 

### PR DESCRIPTION
**Pull request by almilosz**
### Details:
 - Accept sinks as output ports in addition to generic nodes and op class instances in `ov.Model` ctors
 - Add test

 Added support for:
 - `Model(results: List[openvino._pyopenvino.op.Result], sinks: List[ov::Output<ov::Node>], parameters: List[openvino._pyopenvino.op.Parameter], name: str = '')`
 - `Model(results: List[ov::Output<ov::Node>], sinks: List[ov::Output<ov::Node>],  parameters: List[openvino._pyopenvino.op.Parameter], name: str = '')`

### Tickets:
 - *[CVS-131037](https://jira.devtools.intel.com/browse/CVS-131037)*

### CODE REVIEW

File: src/bindings/python/src/pyopenvino/graph/model.cpp [hunk 1]
```diff
@@ -50,6 +50,16 @@ static ov::SinkVector cast_to_sink_vector(const std::vector<std::shared_ptr<ov::
     return sinks;
 }
 
+static ov::SinkVector cast_to_sink_vector(const ov::OutputVector& outputs) {
+    ov::SinkVector sinks;
+    for (const auto& output : outputs) {
+        auto sink = std::dynamic_pointer_cast<ov::op::Sink>(output.get_node_shared_ptr());
+        OPENVINO_ASSERT(sink != nullptr, "Output node handle {} is not instance of Sink");
+        sinks.push_back(sink);
```
Comment (praasz) [inline src/bindings/python/src/pyopenvino/graph/model.cpp:58]:
```suggestion
        sinks.push_back(std::move(sink));
```

File: src/bindings/python/src/pyopenvino/graph/model.cpp [hunk 2]
```diff
@@ -50,6 +50,16 @@ static ov::SinkVector cast_to_sink_vector(const std::vector<std::shared_ptr<ov::
     return sinks;
 }
 
+static ov::SinkVector cast_to_sink_vector(const ov::OutputVector& outputs) {
+    ov::SinkVector sinks;
```
Comment (praasz) [inline src/bindings/python/src/pyopenvino/graph/model.cpp:54]:
```suggestion
    ov::SinkVector sinks;
    sinks.reserve(outputs.size());
```

File: src/bindings/python/src/pyopenvino/graph/model.cpp [hunk 3]
```diff
@@ -50,6 +50,16 @@ static ov::SinkVector cast_to_sink_vector(const std::vector<std::shared_ptr<ov::
...
+    for (const auto& output : outputs) {
+        auto sink = std::dynamic_pointer_cast<ov::op::Sink>(output.get_node_shared_ptr());
+        OPENVINO_ASSERT(sink != nullptr, "Output node handle {} is not instance of Sink");
```
Comment (praasz) [inline src/bindings/python/src/pyopenvino/graph/model.cpp:57]:
```suggestion
        OPENVINO_ASSERT(sink != nullptr, "Output {} is not instance of Sink");
```
?

What `{}` means in error message?

File: src/bindings/python/src/pyopenvino/graph/model.cpp [hunk 4]
```diff
@@ -259,6 +269,65 @@ void regclass_graph_Model(py::module m) {
...
+            :param variables: List of variables.
+            :type variables: List[op.util.Variable]
+            :param name: String to set as model's friendly name.
+            :type name: str
+            )");
+
+    model.def(py::init([](const ov::ResultVector& results,
+                          const ov::OutputVector& nodes,
+                          const ov::ParameterVector& parameters,
+                          const ov::op::util::VariableVector& variables,
```
Comment (akuporos) [inline src/bindings/python/src/pyopenvino/graph/model.cpp:305]: we still need to pass variables to constructor.
If you want to keep this version of constructor where variables can be omitted, may be it is worth to add a check if variablesVector is empty or smthg like this?

**Review by praasz — APPROVED** [review]

File: src/bindings/python/src/pyopenvino/graph/model.cpp [hunk 5]
```diff
@@ -256,9 +266,90 @@ void regclass_graph_Model(py::module m) {
...
+            :type name: str
+            )");
+
+    model.def(py::init([](const ov::OutputVector& results,
+                          const ov::OutputVector& nodes,
+                          const ov::ParameterVector& parameters,
+                          py::object& variables,
+                          const std::string& name) {
+                  set_tensor_names(parameters);
+                  const auto sinks = cast_to_sink_vector(nodes);
```
Comment (akuporos) [inline src/bindings/python/src/pyopenvino/graph/model.cpp:281]: I think it'd be better to split current solution to two ctors: one with variable, other without

File: src/bindings/python/tests/test_runtime/test_model.py [hunk 6]
```diff
@@ -233,6 +233,53 @@ def test_get_sink_index(device):
...
+    # Model(List[openvino._pyopenvino.op.Result], List[ov::Output<ov::Node>],
+    # List[openvino._pyopenvino.op.Parameter], List[openvino._pyopenvino.op.util.Variable], str = '')
+    model = Model(results=[res], sinks=[assign.output(0)], parameters=[input_data], variables=[variable_1], name="TestModel")
+    model.validate_nodes_and_infer_types()
+    assert model.sinks[0].get_output_shape(0) == Shape([2, 2])
+    assert sinks == [sink.get_type_name() for sink in model.get_sinks()]
+
+    # Model(List[ov::Output<ov::Node>, List[ov::Output<ov::Node>],
+    # List[openvino._pyopenvino.op.Parameter], List[openvino._pyopenvino.op.util.Variable], str = '')
+    model = Model(results=[model.output(0)], sinks=[assign.output(0)], parameters=[input_data], variables=[variable_1], name="TestModel")
```
Comment (akuporos) [inline src/bindings/python/tests/test_runtime/test_model.py:277]:
```suggestion
    model = Model(results=[res.output(0)], sinks=[assign.output(0)], parameters=[input_data], variables=[variable_1], name="TestModel")
```

File: src/bindings/python/tests/test_runtime/test_model.py [hunk 7]
```diff
@@ -233,6 +233,53 @@ def test_get_sink_index(device):
...
+    # List[openvino._pyopenvino.op.Parameter], str = '')
+    model = Model(results=[res], sinks=[node.output(0)], parameters=[input_data], name="TestModel")
+    model.validate_nodes_and_infer_types()
+    sinks = ["Assign"]
+    assert sinks == [sink.get_type_name() for sink in model.get_sinks()]
+    assert model.sinks[0].get_output_shape(0) == Shape([2, 2])
+
+    # Model(List[ov::Output<ov::Node>, List[ov::Output<ov::Node>],
+    # List[openvino._pyopenvino.op.Parameter], str = '')
+    model = Model(results=[model.output(0)], sinks=[node.output(0)], parameters=[input_data], name="TestModel")
```
Comment (akuporos) [inline src/bindings/python/tests/test_runtime/test_model.py:253]:
```suggestion
    model = Model(results=[res.output(0)], sinks=[node.output(0)], parameters=[input_data], name="TestModel")
```

File: src/bindings/python/src/pyopenvino/graph/model.cpp (hunk 3 above)
Comment (akuporos) [reply src/bindings/python/src/pyopenvino/graph/model.cpp]: actually nothing, @almilosz removed it

**Review by akuporos — APPROVED** [review]
LGTM,

Thank you for your contribution! :)

**Pull request merged**